src/toil_scripts/adam_uberscript/adam_uberscript.py

```diff
@@ -220,110 +218,179 @@ def get_cluster_size(cluster_name):
+    semaphore = Semaphore.create(domain=domain,
+                                 name=standalone_spark_semaphore_name,
+                                 value=params.spark_sample_slots)
+    while True:
+        with throttle(cluster_scaling_interval_in_seconds):
+            # How many samples are being computed (or waiting to be computed) on the Spark cluster? ...
+            num_unused_sample_slots = semaphore.value
+            num_samples = params.spark_sample_slots - num_unused_sample_slots
+            assert num_samples >= 0
+            if num_samples:
+                # ... At least one sample is. How many Spark workers are needed for all of them.
+                spark_workers_per_sample = params.spark_nodes - 1
+                num_desired_workers = num_samples * spark_workers_per_sample
+                num_workers = len(list_nodes(params.cluster_name, role='spark-slave'))
+                num_masters = len(list_nodes(params.cluster_name, role='spark-master'))
+                if num_masters == 0:
+                    # Absence of master indicates absence of cluster, so create it now.
+                    assert num_workers == 0, 'Orphaned Spark workers detected'
+                    output = check_output(['cgcloud',
+                                           'create-cluster',
+                                           '--list',
+                                           '--zone', aws_region + 'a',
+                                           '--cluster-name', params.cluster_name,
+                                           '--leader-instance-type', params.spark_master_type or params.leader_type,
+                                           '--instance-type', params.spark_instance_type or params.instance_type,
+                                           '--num-workers', str(num_desired_workers),  # TODO: spot options
+                                           '--ssh-opts', '"StrictHostKeyChecking=no"'] +
+                                          role_options(params) +
+                                          ['spark'])
+                    output = parse_cgcloud_list_output(output)
+                    assert len(output) == 1
+                    master = output[0]
+                    assert master['role_name'] == 'spark-master'
+                    SparkMasterAddress.create(domain, master['private_ip_address']).save(force=True)
+                elif num_masters == 1:
+                    # There already is a Spark cluster, so scale it if necessary. 
+                    if num_desired_workers > num_workers:
+                        grow_cluster(num_nodes=num_desired_workers - num_workers,
+                                     instance_type=params.instance_type,
+                                     cluster_name=params.cluster_name,
+                                     cluster_type='spark')
+                    elif num_desired_workers < num_workers:
+                        pass  # TODO: think about scaling down
```

Comment:
To remove instances, sure. But what about the HDFS contents? Don't we need to to some rebalancing before taking instances out of service?